### PR TITLE
Sync IPv6 and Optional DHCP Changes from Upstream

### PR DESCRIPTION
**Pull request by stbenjam**
This PR pulls in the upstream changes:
 - https://github.com/metal3-io/ironic-image/pull/114
 - https://github.com/metal3-io/ironic-image/pull/107

These changes enable IPv6. When $PROVISIONING_IP is specified, which may be an IPv6 address, the various containers will wait for that IP to become available on an interface.

If the IP is IPv6, then we use an IPv6-specific configurations. Note, IPv6 hosts are expected to be using UEFI boot, as we use snponly.efi. snponly.efi uses the UEFI network stack instead of the iPXE drivers. When using EDK2 OVMF + iPXE + ipxe.efi, we have seen lock-ups in initialization of the hardware devices.

As neither CentOS nor RHEL distribute iPXE builds with IPv6 support, we
build them from source as part of the container build.

Additionally, if $DHCP_RANGE is empty, then we don't include the configuration items for DHCP in dnsmasq, only TFTP.

Due to the increased complexity of the dnsmasq.conf configuration, this
also changes the way we template the file to using Jinja2.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,16 +1,32 @@
...
+FROM docker.io/centos:centos7 AS builder
+RUN yum install -y gcc git make genisoimage xz-devel
+WORKDIR /tmp
+COPY . .
+RUN git clone http://git.ipxe.org/ipxe.git && \
+      cd ipxe && \
+      git checkout 3fe683ebab29afacf224e6b0921f6329bebcdca7 && \
+      cd src && \
+      sed -i -e "s/#undef.*NET_PROTO_IPV6/#define NET_PROTO_IPV6/g" config/general.h && \
+      make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi bin-x86_64-efi/snponly.efi
```
Comment (stbenjam) [inline Dockerfile]: https://bugzilla.redhat.com/show_bug.cgi?id=1776929 tracks having this included in RHEL 8. The work is completed, it should show up soon in a RHEL release. https://bugzilla.redhat.com/show_bug.cgi?id=1776929

File: Dockerfile.ocp [hunk 2]
```diff
@@ -13,20 +13,37 @@ RUN if [ $(uname -m) = "x86_64" ]; then \
...
+      make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi bin-x86_64-efi/snponly.efi
+
 FROM ubi8
 
 RUN yum update -y && \
     yum install -y python3-gunicorn openstack-ironic-api openstack-ironic-conductor crudini \
         iproute iptables dnsmasq httpd qemu-img parted gdisk ipxe-bootimgs psmisc procps-ng \
-        mariadb-server ipxe-roms-qemu genisoimage python3-ironic-prometheus-exporter && \
+        mariadb-server ipxe-roms-qemu genisoimage python3-ironic-prometheus-exporter \
+        python-jinja2 && \
```
Comment (hardys) [inline Dockerfile.ocp]: Should be python3-jinja2 I think?

File: Dockerfile.ocp [hunk 3]
```diff
@@ -13,20 +13,37 @@ RUN if [ $(uname -m) = "x86_64" ]; then \
       touch /tmp/esp.img; \
     fi
 
+## Build iPXE w/ IPv6 Support
+## Note: we are pinning to a specific commit for reproducible builds.
+## Updated as needed.
+RUN git clone http://git.ipxe.org/ipxe.git && \
```
Comment (elfosardo) [inline Dockerfile.ocp]: I'm pretty sure this is not accessible downstream :/

Comment (stbenjam) [reply Dockerfile.ocp]: It's accessible, although certainly not ideal. We'll have the right things in RPM's soon(ish) https://github.com/openshift/ironic-image/pull/45#discussion_r365261838

**Comment by elfosardo** [comment]
@stbenjam can you please add the link to the upstream change in the description?
Also I think changes to Dockerfile are not needed since this image doesn't use it.

**Comment by stbenjam** [comment]
> @stbenjam can you please add the link to the upstream change in the description?
> Also I think changes to Dockerfile are not needed since this image doesn't use it.

Done. Is there a way to get the image from CI? I can't build it locally since the openshift images use special yum repos:

```
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"
```

**Comment by elfosardo** [comment]
@stbenjam there's a way to build the image locally, but it's not straightforward, you will have to add the correct repositories in the Dockerfile.ocp.

In case you want to get the image built from CI, check the build logs and you'll see lines like:
"images will be pullable from registry.svc.ci.openshift.org/ci-op-xc60624s/stable:${component}"
"Create release image registry.svc.ci.openshift.org/ci-op-xc60624s/release:latest"
For example for this job:
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_ironic-image/45/pull-ci-openshift-ironic-image-master-images/96/build-log.txt

**Comment by stbenjam** [comment]
I'm able to test this, and looks like it's all working.

We need the installer changes in https://github.com/openshift/installer/pull/2895 to land first, as we have to set DHCP_RANGE now, there's no default value (as blank now will mean disabled).

**Comment by stbenjam** [comment]
https://github.com/openshift/installer/pull/2895 merged


/hold cancel

**Pull request merged**